### PR TITLE
feat: add DynamicDutyCalculator

### PR DESCRIPTION
**Pull request by razww**
This PR implement dynamic interest rate mechanism which calculate interest rate based on lisUSD price:
```
power = (peg - price) / beta
rate = rate0 * e^power
```
in which `rate0` and `beta` are pre-defined value

### CODE REVIEW

File: contracts/Interaction.sol [hunk 1]
```diff
@@ -326,7 +329,14 @@ contract Interaction is OwnableUpgradeable, IDao, IAuctionProxy {
         CollateralType memory collateralType = collaterals[token];
         _checkIsLive(collateralType.live);
 
-        jug.drip(collateralType.ilk);
+        bytes32 _ilk = collateralType.ilk;
+        (uint256 currentDuty,) = jug.ilks(_ilk);
+        uint256 duty = dutyCalculator.calculateDuty(token, currentDuty, true);
+        if (duty != currentDuty) {
+            setCollateralDuty(token, duty);
```
Comment (razww) [inline contracts/Interaction.sol:336]: This is wrong since `msg.sender` can be any address which may not authorized.

**Review by qingyang-lista — DISMISSED** [review]
LGTM

File: contracts/Interaction.sol (hunk 1 above)
Comment (razww) [reply contracts/Interaction.sol]: Fixed in `81d604dfc3bc90e2965c186cabc69d7d9977064a `

File: contracts/amo/DynamicDutyCalculator.sol [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        emit DutyRangeUpdated(_minDuty, _maxDuty);
+    }
+
+    /**
+     * @dev Set the price deviation required between two consecutive duty updates.
+     * @param _priceDeviation The price deviation required between two consecutive duty updates.
+     */
+    function setPriceDeviation(uint256 _priceDeviation) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(priceDeviation != _priceDeviation && _priceDeviation <= minPrice, "AggMonetaryPolicy/invalid-price-deviation");
+        require(priceDeviation < (maxPrice - minPrice), "AggMonetaryPolicy/priceDeviation-is-too-large");
```
Comment (razww) [inline contracts/amo/DynamicDutyCalculator.sol:211]: should be `_priceDeviation < (maxPrice - minPrice)`

File: contracts/amo/DynamicDutyCalculator.sol [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        lisUSD = _lisUSD;
+        oracle = IResilientOracle(_priceOracle);
+
+        minDuty = 1e27;
+        maxDuty = 1000000034836767751273470154;
+
+        minPrice = 9e7;
+        maxPrice = 11e7;
+
+        require(priceDeviation < (maxPrice - minPrice), "AggMonetaryPolicy/invalid-price-deviation");
```
Comment (razww) [inline contracts/amo/DynamicDutyCalculator.sol:91]: should be `_priceDeviation < (maxPrice - minPrice)`

Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol]: Fixed in commit 7b4d220acc09d1cbfecb7b9164d7de58634d36f2

File: contracts/Interaction.sol [hunk 4]
```diff
@@ -593,4 +607,18 @@ contract Interaction is OwnableUpgradeable, IDao, IAuctionProxy {
...
         require(live != 0, "Interaction/inactive collateral");
     }
+
+    function setDutyCalculator(address _dutyCalculator) external auth {
+        require(_dutyCalculator != address(0) && _dutyCalculator != address(dutyCalculator), "Interaction/invalid-dutyCalculator-address");
+        dutyCalculator = IDynamicDutyCalculator(_dutyCalculator);
+        require(dutyCalculator.interaction() == address(this), "Interaction/invalid-dutyCalculator-interaction");
+    }
+
+    function getNextDuty(address _collateral) external returns (uint256 duty) {
```
Comment (razww) [inline contracts/Interaction.sol:617]: Need tests to ensure this method can be invoked as `view`

Comment (razww) [reply contracts/Interaction.sol:623]: Added in b5770a310859608f177afa392a0686020c277210

File: contracts/amo/DynamicDutyCalculator.sol [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+        interaction = _interaction;
+        lisUSD = _lisUSD;
+        oracle = IResilientOracle(_priceOracle);
+
+        minDuty = 1e27;
+        maxDuty = 1000000034836767751273470154;
+
+        minPrice = 9e7;
+        maxPrice = 11e7;
```
Comment (razww) [inline contracts/amo/DynamicDutyCalculator.sol:89]: better to be in forms of PEG factors

File: contracts/amo/DynamicDutyCalculator.sol [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    address public interaction;
+
+    // lisUSD address
+    address lisUSD;
+
+    // lisUSD price oracle
+    IResilientOracle oracle;
+
+    // the minimum price deviation required between two consecutive duty updates
+    uint256 public priceDeviation;
```
Comment (razww) [inline contracts/amo/DynamicDutyCalculator.sol:41]: add cap for `priceDeviation`

Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol:41]: rename `priceDeviation` to `delta`

File: contracts/amo/DynamicDutyCalculator.sol [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+    /**
+     * @dev Adaptor for `FixedMath0x._exp` method since it's designed for negative power only.
+     *      if price > 1, use _exp directly
+     *      if price < 1, e^power = 1 / e^(-power)
+     * @param delta  (peg price - lisUSD price)
+     * @param beta  volatility parameter; initial value is 1e6
+     * @return e^(delta/beta)
+     */
+    function exp(int256 delta, int256 beta) internal pure returns (uint256) {
```
Comment (razww) [inline contracts/amo/DynamicDutyCalculator.sol:255]: rename `delta` to `deviation`

File: contracts/amo/DynamicDutyCalculator.sol [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+import "../oracle/interfaces/IResilientOracle.sol";
+
+/**
+    * @title DynamicDutyCalculator
+    * @author Lista
+    * @notice Contract for calculating the dynamic interest rate for collaterals to maintain the stability of the lisUSD peg.
+    *
+    * The calculation is based on the following formula:
+    *
+    *   delta = PEG - lisUSD price
```
Comment (razww) [inline contracts/amo/DynamicDutyCalculator.sol:19]: `delta` -> `deviation`

File: contracts/amo/DynamicDutyCalculator.sol [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    /**
+     * @dev Calculate rate given the lisUSD price, beta, and rate0.
+     * @param price The lisUSD price.
+     * @param beta The volatility parameter.
+     * @param rate0 The rate when the price is equal to PEG.
+     */
+    function calculateRate(uint256 price, uint256 beta, uint256 rate0) internal pure returns (uint256 rate) {
+        int256 delta = int256(PEG) - int256(price);
```
Comment (razww) [inline contracts/amo/DynamicDutyCalculator.sol:171]: `delta` -> `deviation`

File: contracts/amo/DynamicDutyCalculator.sol (hunk 5 above)
Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol]: done in 795a7e45157fe175da9def0e4b51369fda63400c

File: contracts/amo/DynamicDutyCalculator.sol (hunk 6 above)
Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol]: > rename `priceDeviation` to `delta`

done in 7ae7cffd648d7848e39f710405f85c26af0b820c

File: contracts/amo/DynamicDutyCalculator.sol (hunk 7 above)
Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol]: done in 7ae7cffd648d7848e39f710405f85c26af0b820c

File: contracts/amo/DynamicDutyCalculator.sol (hunk 9 above)
Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol]: done in 7ae7cffd648d7848e39f710405f85c26af0b820c

File: contracts/amo/DynamicDutyCalculator.sol (hunk 8 above)
Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol]: done in 7ae7cffd648d7848e39f710405f85c26af0b820c

File: contracts/amo/DynamicDutyCalculator.sol (hunk 6 above)
Comment (razww) [reply contracts/amo/DynamicDutyCalculator.sol]: > add cap for `priceDeviation`

added cap for duty instead 8b83905c45463324dd016492c8aad7851e2f8011

**Review by qingyang-lista — APPROVED** [review]
LGTM

**Review by qingyang-lista — APPROVED** [review]
LGTM

**Pull request merged**